src/app/features/search-bar/search-bar.component.ts

```diff
@@ -0,0 +1,250 @@
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class SearchBarComponent implements AfterViewInit, OnDestroy {
+  @Input() isElevated: boolean = false;
+  @Output() blurred: EventEmitter<any> = new EventEmitter();
+
+  @ViewChild('inputEl') inputEl!: ElementRef;
+  @ViewChild('searchForm') searchForm!: ElementRef;
+
+  T: typeof T = T;
+  isLoading$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);
+  taskSuggestionsCtrl: FormControl = new FormControl();
+  filteredIssueSuggestions$: Observable<SearchItem[]> = new Observable();
+  isArchivedTasks: boolean = false;
+  tooManyResults: boolean = false;
+
+  private _attachKeyDownHandlerTimeout?: number;
+  private _tasks: Task[] = [];
+  private _projects: Project[] = [];
+  private _tags: Tag[] = [];
+  private _searchableItems: SearchItem[] = [];
+
+  constructor(
+    private _taskService: TaskService,
+    private _projectService: ProjectService,
+    private _tagService: TagService,
+    private _router: Router,
+  ) {
+    this._taskService.allTasks$.pipe(take(1)).subscribe((t) => (this._tasks = t));
+    this._projectService.list$.pipe(take(1)).subscribe((p) => (this._projects = p));
+    this._tagService.tags$.pipe(take(1)).subscribe((t) => (this._tags = t));
+
+    this._mapTasksToSearchItems();
+  }
+
+  private _mapTasksToSearchItems(): void {
+    this._searchableItems = this._tasks.map((task) => {
+      const item: SearchItem = {
+        id: task.id,
+        title: task.title.toLowerCase(),
+        taskNotes: task.notes.toLowerCase(),
+        projectId: task.projectId,
+        parentId: task.parentId,
+        tagIds: task.tagIds,
+        timeSpentOnDay: task.timeSpentOnDay,
+        createdOn: task.created,
+        issueType: task.issueType,
+        ctx: this._getContextIcon(task),
+      };
+      return item;
+    });
+  }
+
+  private _getContextIcon(task: Task) {
+    const context = this._getCtxForTaskSuggestion(task);
+    return {
+      ...context,
+      icon: (context as Tag).icon || (task.projectId && 'list'),
+    };
+  }
+
+  ngAfterViewInit(): void {
+    this._setUpFilter();
+
+    this._attachKeyDownHandlerTimeout = window.setTimeout(() => {
+      this.inputEl.nativeElement.addEventListener('keydown', (ev: KeyboardEvent) => {
+        if (ev.key === 'Escape') {
+          this.blurred.emit();
+        } else if (ev.key === '1' && ev.ctrlKey) {
+          this.switchTaskSource();
+          ev.preventDefault();
+        } else if (
+          ev.key === 'Enter' &&
+          (!this.taskSuggestionsCtrl.value ||
+            typeof this.taskSuggestionsCtrl.value === 'string')
+        ) {
+          this._shakeSearchForm();
+          ev.preventDefault();
+        }
+      });
+    });
+  }
+
+  private _triggerFilter(): void {
+    this.taskSuggestionsCtrl.setValue(this.inputEl.nativeElement.value);
+  }
+
+  private _setUpFilter(): void {
+    this.filteredIssueSuggestions$ = this.taskSuggestionsCtrl.valueChanges.pipe(
```

Comment:
This is a bit tricky and might lead to errors later on, as subscriptions to any previously assigned values might live on. Generally, I personally prefer to assign Observables once initially and never reassign them again.

Reply:
Ok good to know. Thanks for explaining this.
I don't even remember why I put it in it's own method as it is only called that one time in the beginning.
But at least I learned something :)